share/man/man7/release.7
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2002 Murray Stokely <murray@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 13, 2025
.Dt RELEASE 7
.Os
.Sh NAME
.Nm release
.Nd "release building infrastructure"
.Sh DESCRIPTION
.Fx
provides a complete build environment suitable for users to make
full releases of the
.Fx
operating system.
All of the tools necessary to build a release are available from the
.Fx
source code repository in
.Pa src/release .
A complete release can be built with only a single command,
including the creation of ISO images suitable for burning to CD-ROM,
memory stick images, and a network install directory.
This command is aptly named
.Dq Li "make release" .
.Pp
For some users, it may be desirable to provide an absolutely clean
build environment, with no local modifications to the source tree or to
.Xr make.conf 5 ,
and with clean checkouts of specific versions of the doc, src, and ports
trees.
For this purpose, a script
.Pq Pa src/release/release.sh
is provided to automate these checkouts and then execute
.Dq Li "make release"
in a clean
.Xr chroot 8 .
.Pp
Before attempting to build a release, the user is expected to be
familiar with the contents of
.Xr build 7 ,
and should have experience upgrading systems from source.
.Pp
The release build process requires that
.Pa /usr/obj
be populated with the output of
.Dq Li "make buildworld"
and
.Dq Li "make buildkernel" .
This is necessary to provide the object files for the release or, when
using
.Pa release.sh ,
so that the object files for a complete system can be installed into a clean
.Xr chroot 8
environment.
.Pp
If the target release build is for a different architecture or machine type,
the
.Va TARGET
and
.Va TARGET_ARCH
variables must be used.
See the supported
.Fa release.conf
variables for more information.
.Pp
The release procedure on some architectures may also require that the
.Xr md 4
(memory disk) device driver be present in the kernel
.Pq either by being compiled in or available as a module .
.Pp
This document does not cover source code management, quality
assurance, or other aspects of the release engineering process.
.Sh CLEAN RELEASE GENERATION
Official releases of
.Fx
are produced in a clean environment to
ensure consistency between the versions of the src, ports, and doc trees
and to avoid contamination from the host system
.Po such as local patches, changes
to
.Xr make.conf 5 ,
etc.
.Pc .
This is accomplished using the wrapper script
.Pa src/release/release.sh .
.Pp
.Ic release.sh
.Op Fl c Ar release.conf
.Pp
.Ic release.sh
checks out the
.Li src/ ,
.Li ports/ ,
and
.Li doc/
trees to
.Va CHROOTDIR ,
then calls
.Dq Li "make buildworld"
and
.Dq Li "make installworld"
to generate a
.Xr chroot 8
environment.
Next,
.Dq Li "make release"
is run within the
.Xr chroot 8
environment and places the result in
.Pa $CHROOTDIR/R .
.Pp
The optional
.Fa release.conf
configuration file supports the following variables:
.Bl -tag -width Ev
.It Va CHROOTDIR
The directory within which the release will be built.
Defaults to
.Pa /scratch .
.It Va CHROOT_MAKEENV
Additional
.Xr make 1
arguments to pass through, which directly affect the
tuning of the build chroot.
.It Va NOGIT
Do not explicitly require the
.Xr git 1
port to be installed.
.It Va SRCGITROOT
The
.Xr git 1
host used to check out the 5BSD source tree.
Defaults to
.Pa https://github.com/5BSD/ .
.It Va PORTGITROOT
The
.Xr git 1
host used to check out the upstream ports tree.
Defaults to
.Pa https://git.FreeBSD.org/ .
.It Va SRCBRANCH
The
.Li src/
branch to use.
Defaults to
.Fl b Va main .
.It Va PORTBRANCH
The
.Li ports/
branch to use.
Defaults to
.Va head/@rHEAD .
.It Va TARGET
The target machine type for cross-building a release.
.It Va TARGET_ARCH
The target machine architecture for cross-building a release.
.Pp
For the supported list of
.Va TARGET
and
.Va TARGET_ARCH
combinations, consult the output of
.Dq make targets
as documented in
.Xr build 7 .
.It Va KERNEL
The target kernel configuration to use.
Defaults to
.Va GENERIC .
Multiple
.Va KERNEL
entries may be specified.
.It Va MAKE_CONF
The
.Xr make.conf 5
to use for the release build.
Defaults to
.Fa /dev/null
to prevent polluting the release with local system changes.
.It Va SRC_CONF
The
.Xr src.conf 5
to use for the release build.
Defaults to
.Fa /dev/null
to prevent polluting the release with local system changes.
.It Va MAKE_FLAGS
Additional flags to pass to
.Xr make 1 .
.It Va WORLD_FLAGS
Additional flags to pass to
.Xr make 1
during the
.Dq buildworld
phase.
Defaults to setting the number of
.Xr make 1
jobs
.Pq Ar -j
to the number of CPUs available on a SMP-capable system.
.It Va KERNEL_FLAGS
Additional flags to pass to
.Xr make 1
during the
.Dq buildkernel
phase.
Defaults to setting the number of
.Xr make 1
jobs
.Pq Ar -j
to half the number of CPUs available on a SMP-capable system.
.It Va NOPORTS
Set to a non-empty value to skip the
.Li ports/
tree checkout.
When set,
.Va NOPORTS
will prevent the
.Fa ports.txz
distribution package from being created.
.It Va WITH_DVD
Set to a non-empty value to include the
.Cm dvdrom
target.
.It Va WITH_COMPRESSED_IMAGES
Set to a non-empty value to compress the release images with
.Xr xz 1 .
The original
.Pq uncompressed
images are not removed.
.It Va XZ_THREADS Pq Vt int
Set to the number of threads
.Xr xz 1
should use when compressing images.
By default,
.Va XZ_THREADS
is set to
.Va 0 ,
which uses all available cores on the system.
.It Va VCSCMD
The command run to obtain the source trees.
Defaults to
.Qq Cm git clone Fl q .
.It Va CHROOTBUILD_SKIP
If defined, the
.Li buildworld ,
.Li installworld ,
and
.Li distribution
stages of the
.Xr chroot 8
build environment setup are skipped.
This is intended solely for cases where the
.Xr chroot 8
userland are provided by alternate means.
.It Va SRC_UPDATE_SKIP
Set to a non-empty value to prevent checkout or update of
.Fa /usr/src
within the
.Xr chroot 8 .
This is intended for use only when
.Fa /usr/src
is expected to exist by alternative means.
.It Va PORTS_UPDATE_SKIP
Set to a non-empty value to prevent checkout or update of
.Fa /usr/ports
within the
.Xr chroot 8 .
This is intended for use only when
.Fa /usr/ports
is expected to exist by alternative means.
.It Va NOPKGBASE
Include legacy tarball distribution sets for use on the install media,
instead of base system packages.
.It Va PKG_CMD
A path to the
.Xr pkg 8
executable to use when installing packages in release images as a non-root user.
.It Va PKG_REPOS_DIR
An optional path to a directory containing
.Xr pkg 8
repository configuration files.
These configuration files will be used when installing packages in release
images as a non-root user.
.It Va PKG_REPO_NAME
The name of the repository configuration to use when installing packages in
release images as a non-root user.
.El
.Sh EMBEDDED BUILDS
The following
.Fa release.conf
variables are relevant only to release builds for embedded systems:
.Bl -tag -width Ev
.It Va EMBEDDEDBUILD
Set to a non-null value to enable functionality for embedded device
release builds.
.Pp
When set,
.Va WITH_DVD
is unset.
Additionally,
.Va EMBEDDED_TARGET
and
.Va EMBEDDED_TARGET_ARCH
must also be defined.
When the build environment is created,
.Fa release.sh
runs a separate build script located in an architecture-specific
directory in
.Pa src/release/${EMBEDDED_TARGET}/ .
.It Va EMBEDDEDPORTS
Set to the list of any ports that are required for the target device
in the format of
.Fa category/port .
.It Va EMBEDDED_TARGET
When set, its value is passed to
.Xr make 1
to set the
.Va TARGET
.Pq value of Cm uname Fl m
to cross build the target userland.
.It Va EMBEDDED_TARGET_ARCH
When set, its value is passed to
.Xr make 1
to set the
.Va TARGET_ARCH
.Pq value of Cm uname Fl p
to cross build the target userland.
.El
.Sh VIRTUAL MACHINE DISK IMAGES
The following
.Fa release.conf
variables are relevant only to virtual machine disk image builds:
.Bl -tag -width Ev
.It Va WITH_VMIMAGES
Set to a non-null value to build virtual machine disk images as part
of the release build.
.Va WITH_VMIMAGES
may also be specified as an environment variable passed to
.Xr make 1 .
.It Va WITH_COMPRESSED_VMIMAGES
Set to a non-null value to compress the virtual machine disk images with
.Xr xz 1
as part of the
.Cm install
.Xr make 1
target.
Note that compressing virtual machine disk images may take a very long
time on some systems.
.It Va VMBASE
Set to change the name of the resulting virtual machine disk image file.
The default value is
.Va vm .
.It Va VMSIZE
Set to change the size of the virtual machine disk capacity.
The default value is
.Va 20g .
See
.Xr makefs 8
for valid values.
.Pp
Virtual machine disk images are, by default, created as sparse images.
When
.Va WITH_COMPRESSED_VMIMAGES
is used, the resulting files compressed with
.Xr xz 1
compress to roughly the same size, regardless of the specified disk image
size.
.It Va VMFS
(Deprecated.)
Set to specify which of the filesystem(s) listed in
.Va VMFSLIST
is linked to the historical non-filesystem-labelled file name.
Valid values are
.Va ufs
and
.Va zfs .
The default value is
.Va ufs .
.It Va VMFSLIST
Set to specify the list of file system types to build images for.
Valid values are one or both of
.Va ufs
and
.Va zfs .
The default value is
.Va ufs zfs .
.It Va VMFORMATS
Set to the target virtual disk image format(s) to create.
By default, the
.Va vhdf , Va vmdk , Va qcow2 ,
and
.Va raw
formats are created.
See
.Xr mkimg 1
for valid format values.
.El
.Pp
For a list of supported
.Va VMFORMATS
values
.Pq including cloud hosting provider formats
along with a brief description, run:
.Bd -literal -offset indent
cd /usr/src
make -C release list-vmtargets
.Ed
.Sh CLOUD HOSTING MACHINE IMAGES
The
.Fx
release build tools support building virtual machine images for various
cloud hosting providers, each with their own specific configuration to
include support for each hosting provider by default.
.Pp
The following
.Xr make 1

release/release.sh
#!/bin/sh
#-
# Copyright (c) 2020-2021 Rubicon Communications, LLC (netgate.com)
# Copyright (c) 2013-2019 The FreeBSD Foundation
# Copyright (c) 2013 Glen Barber
# Copyright (c) 2011 Nathan Whitehorn
# All rights reserved.
#
# Portions of this software were developed by Glen Barber
# under sponsorship from the FreeBSD Foundation.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
# release.sh: check out source trees, and build release components with
#  totally clean, fresh trees.
# Based on release/generate-release.sh written by Nathan Whitehorn
#

export PATH="/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin"

VERSION=3

# Prototypes that can be redefined per-chroot or per-target.

# load_chroot_env(): Set up the build environment needed.
#
# Done as part of chroot_env().
load_chroot_env() { }

# load_target_env(): set up the build environment needed for the
# chroot_build_target() and `${chroot_build_release}` steps.
load_target_env() { }

# buildenv_setup(): set up the build environment needed for post-chroot_setup()
buildenv_setup() { }

# chroot_cleanup(): Clean up resources setup in chroot_setup() at exit.
#
# This function can be built upon. `_chroot_cleanup` must be added to the end of
# the override function, if overridden.
chroot_cleanup() {
	_chroot_cleanup
} # chroot_cleanup()

usage() {
	echo "Usage: $0 [-c release.conf]"
	exit 1
}

# env_setup(): Set up the default build environment variables, such as the
# CHROOTDIR, VCSCMD, SRCGITROOT, etc.  This is called before the release.conf
# file is sourced, if '-c <release.conf>' is specified.
env_setup() {
	# The directory within which the release will be built.
	CHROOTDIR="/scratch"
	if [ -z "${RELENGDIR}" ]; then
		export RELENGDIR="$(dirname $(realpath ${0}))"
	fi

	# The default version control system command to obtain the sources.
	for _dir in /usr/bin /usr/local/bin; do
		[ -x "${_dir}/git" ] && VCSCMD="/${_dir}/git"
		[ -n "${VCSCMD}" ] && break 2
	done

	if [ -z "${VCSCMD}" -a -z "${NOGIT}" ]; then
		echo "*** The devel/git port/package is required."
		exit 1
	fi
	VCSCMD="/usr/local/bin/git clone -q"

	# The default source and ports checkout servers and branches.
	SRCGITROOT="https://github.com/5BSD/"
	PORTGITROOT="https://git.FreeBSD.org/"
	SRCBRANCH="main"
	PORTBRANCH="main"
	GITSRC="5BSD.git"
	GITPORTS="ports.git"

	# Set for embedded device builds.
	EMBEDDEDBUILD=

	# The default make.conf and src.conf to use.  Set to /dev/null
	# by default to avoid polluting the chroot(8) environment with
	# non-default settings.
	MAKE_CONF="/dev/null"
	SRC_CONF="/dev/null"

	# The number of make(1) jobs, defaults to the number of CPUs available
	# for buildworld, and half of number of CPUs available for buildkernel
	# and 'make release'.
	WORLD_FLAGS="-j$(sysctl -n hw.ncpu)"
	KERNEL_FLAGS="-j$(( $(( $(sysctl -n hw.ncpu) + 1 )) / 2))"
	RELEASE_FLAGS="-j$(( $(( $(sysctl -n hw.ncpu) + 1 )) / 2))"

	MAKE_FLAGS="-s"

	# The name of the kernel to build, defaults to 5BSD.
	KERNEL="VBSD"

	# Set to non-empty value to disable checkout of doc/ and/or ports/.
	NOPORTS=

	# Set to non-empty value to disable distributing source tree.
	NOSRC=

	# Set to non-empty value to build dvd1.iso as part of the release.
	WITH_DVD=
	WITH_COMPRESSED_IMAGES=

	# Set to non-empty value to build virtual machine images as part of
	# the release.
	WITH_VMIMAGES=
	WITH_COMPRESSED_VMIMAGES=
	XZ_THREADS=0

	# Set to non-empty value to build virtual machine images for various
	# cloud providers as part of the release.
	WITH_CLOUDWARE=

	# Set to non-empty to build OCI images as part of the release
	WITH_OCIIMAGES=

	return 0
} # env_setup()

# env_check(): Perform sanity tests on the build environment, such as ensuring
# files/directories exist, as well as adding backwards-compatibility hacks if
# necessary.  This is called unconditionally, and overrides the defaults set
# in env_setup() if '-c <release.conf>' is specified.
env_check() {
	chroot_build_release_cmd="chroot_build_release"

	# Prefix the repositories with their roots for the full checkout URLs.
	SRC="${SRCGITROOT}${GITSRC}"
	PORT="${PORTGITROOT}${GITPORTS}"

	if [ -n "${EMBEDDEDBUILD}" ]; then
		WITH_DVD=
		WITH_COMPRESSED_IMAGES=
		case ${EMBEDDED_TARGET}:${EMBEDDED_TARGET_ARCH} in
			arm:arm*|arm64:aarch64|riscv:riscv64*)
				chroot_build_release_cmd="chroot_arm_build_release"
				;;
			*)
				;;
		esac
	fi

	# If NOSRC and/or NOPORTS are unset, they must not pass to make
	# as variables.  The release makefile verifies definedness of the
	# NOPORTS variable instead of its value.
	SRCPORTS=
	if [ -n "${NOPORTS}" ]; then
		SRCPORTS="NOPORTS=yes"
	fi
	if [ -n "${NOSRC}" ]; then
		SRCPORTS="${SRCPORTS}${SRCPORTS:+ }NOSRC=yes"
	fi

	# The aggregated build-time flags based upon variables defined within
	# this file, unless overridden by release.conf.  In most cases, these
	# will not need to be changed.
	CONF_FILES="__MAKE_CONF=${MAKE_CONF} SRCCONF=${SRC_CONF}"
	NOCONF_FILES="__MAKE_CONF=/dev/null SRCCONF=/dev/null"
	if [ -n "${TARGET}" ] && [ -n "${TARGET_ARCH}" ]; then
		ARCH_FLAGS="TARGET=${TARGET} TARGET_ARCH=${TARGET_ARCH}"
	else
		ARCH_FLAGS=
	fi

	if [ -z "${CHROOTDIR}" ]; then
		echo "Please set CHROOTDIR."
		exit 1
	fi

	if [ $(id -u) -ne 0 ]; then
		echo "Needs to be run as root."
		exit 1
	fi

	# Unset CHROOTBUILD_SKIP if the chroot(8) does not appear to exist.
	if [ -n "${CHROOTBUILD_SKIP}" -a ! -e ${CHROOTDIR}/bin/sh ]; then
		CHROOTBUILD_SKIP=
	fi

	CHROOT_MAKEENV="${CHROOT_MAKEENV} \
		MAKEOBJDIRPREFIX=${CHROOTDIR}/tmp/obj"
	CHROOT_WMAKEFLAGS="${MAKE_FLAGS} ${WORLD_FLAGS} ${NOCONF_FILES}"
	CHROOT_IMAKEFLAGS="${WORLD_FLAGS} ${NOCONF_FILES}"
	CHROOT_DMAKEFLAGS="${WORLD_FLAGS} ${NOCONF_FILES}"
	RELEASE_WMAKEFLAGS="${MAKE_FLAGS} ${WORLD_FLAGS} ${ARCH_FLAGS} \
		${CONF_FILES}"
	RELEASE_KMAKEFLAGS="${MAKE_FLAGS} ${KERNEL_FLAGS} \
		KERNCONF=\"${KERNEL}\" ${ARCH_FLAGS} ${CONF_FILES}"
	RELEASE_RMAKEFLAGS="${ARCH_FLAGS} ${RELEASE_FLAGS} \
		KERNCONF=\"${KERNEL}\" ${CONF_FILES} ${SRCPORTS} \
		WITH_DVD=${WITH_DVD} WITH_VMIMAGES=${WITH_VMIMAGES} \
		WITH_CLOUDWARE=${WITH_CLOUDWARE} WITH_OCIIMAGES=${WITH_OCIIMAGES} \
		XZ_THREADS=${XZ_THREADS} NOPKGBASE=${NOPKGBASE}"
	RELEASE_RMAKEFLAGS="${RELEASE_RMAKEFLAGS} NO_ROOT=1 WITHOUT_QEMU=1"

	return 0
} # env_check()

# chroot_setup(): Prepare the build chroot environment for the release build.
chroot_setup() {
	load_chroot_env
	mkdir -p ${CHROOTDIR}/usr

	if [ -z "${SRC_UPDATE_SKIP}" ]; then
		if [ -d "${CHROOTDIR}/usr/src/.git" ]; then
			git -C ${CHROOTDIR}/usr/src pull -q
		else
			${VCSCMD} ${SRC} -b ${SRCBRANCH} ${CHROOTDIR}/usr/src
		fi
	fi
	if [ -z "${NOPORTS}" ] && [ -z "${PORTS_UPDATE_SKIP}" ]; then
		if [ -d "${CHROOTDIR}/usr/ports/.git" ]; then
			git -C ${CHROOTDIR}/usr/ports pull -q
		else
			${VCSCMD} ${PORT} -b ${PORTBRANCH} ${CHROOTDIR}/usr/ports
		fi
	fi

	if [ -z "${CHROOTBUILD_SKIP}" ]; then
		cd ${CHROOTDIR}/usr/src
		env ${CHROOT_MAKEENV} make ${CHROOT_WMAKEFLAGS} buildworld
		env ${CHROOT_MAKEENV} make ${CHROOT_IMAKEFLAGS} installworld \
			DESTDIR=${CHROOTDIR}
		env ${CHROOT_MAKEENV} make ${CHROOT_DMAKEFLAGS} distribution \
			DESTDIR=${CHROOTDIR}
	fi

	return 0
} # chroot_setup()

# extra_chroot_setup(): Prepare anything additional within the build
# necessary for the release build.
extra_chroot_setup() {
	mkdir -p ${CHROOTDIR}/dev
	mount -t devfs devfs ${CHROOTDIR}/dev
	[ -e /etc/resolv.conf -a ! -e ${CHROOTDIR}/etc/resolv.conf ] && \
		cp /etc/resolv.conf ${CHROOTDIR}/etc/resolv.conf
	# Run ldconfig(8) in the chroot directory so /var/run/ld-elf*.so.hints
	# is created.  This is needed by ports-mgmt/pkg.
	eval chroot ${CHROOTDIR} /etc/rc.d/ldconfig forcerestart

	# If MAKE_CONF and/or SRC_CONF are set and not character devices
	# (/dev/null), copy them to the chroot.
	if [ -e ${MAKE_CONF} ] && [ ! -c ${MAKE_CONF} ]; then
		mkdir -p ${CHROOTDIR}/$(dirname ${MAKE_CONF})
		cp ${MAKE_CONF} ${CHROOTDIR}/${MAKE_CONF}
	fi
	if [ -e ${SRC_CONF} ] && [ ! -c ${SRC_CONF} ]; then
		mkdir -p ${CHROOTDIR}/$(dirname ${SRC_CONF})
		cp ${SRC_CONF} ${CHROOTDIR}/${SRC_CONF}
	fi

	_gitcmd="$(which git)"
	if [ -z "${NOGIT}" -a -z "${_gitcmd}" ]; then
		# Install git from ports if the ports tree is available;
		# otherwise install the pkg.
		if [ -d ${CHROOTDIR}/usr/ports ]; then
			# Trick the ports 'run-autotools-fixup' target to do the right
			# thing.
			_OSVERSION=$(chroot ${CHROOTDIR} /usr/bin/uname -U)
			REVISION=$(chroot ${CHROOTDIR} make -C /usr/src/release -V REVISION)
			BRANCH=$(chroot ${CHROOTDIR} make -C /usr/src/release -V BRANCH)
			UNAME_r=${REVISION}-${BRANCH}
			GITUNSETOPTS="CONTRIB CURL CVS GITWEB GUI HTMLDOCS"
			GITUNSETOPTS="${GITUNSETOPTS} ICONV NLS P4 PERL"
			GITUNSETOPTS="${GITUNSETOPTS} SEND_EMAIL SUBTREE SVN"
			GITUNSETOPTS="${GITUNSETOPTS} PCRE PCRE2"
			PBUILD_FLAGS="OSVERSION=${_OSVERSION} BATCH=yes"
			PBUILD_FLAGS="${PBUILD_FLAGS} UNAME_r=${UNAME_r}"
			PBUILD_FLAGS="${PBUILD_FLAGS} OSREL=${REVISION}"
			PBUILD_FLAGS="${PBUILD_FLAGS} WRKDIRPREFIX=/tmp/ports"
			PBUILD_FLAGS="${PBUILD_FLAGS} DISTDIR=/tmp/distfiles"
			eval chroot ${CHROOTDIR} env OPTIONS_UNSET=\"${GITUNSETOPTS}\" \
				${PBUILD_FLAGS} \
				make -C /usr/ports/devel/git FORCE_PKG_REGISTER=1 \
				WRKDIRPREFIX=/tmp/ports \
				DISTDIR=/tmp/distfiles \
				install clean distclean
		else
			eval chroot ${CHROOTDIR} env ASSUME_ALWAYS_YES=yes \
				pkg install -y devel/git
			eval chroot ${CHROOTDIR} env ASSUME_ALWAYS_YES=yes \
				pkg clean -y
		fi
	fi

	if [ -n "${EMBEDDEDPORTS}" ]; then
		_OSVERSION=$(chroot ${CHROOTDIR} /usr/bin/uname -U)
		REVISION=$(chroot ${CHROOTDIR} make -C /usr/src/release -V REVISION)
		BRANCH=$(chroot ${CHROOTDIR} make -C /usr/src/release -V BRANCH)
		UNAME_r=${REVISION}-${BRANCH}
		PBUILD_FLAGS="OSVERSION=${_OSVERSION} BATCH=yes"
		PBUILD_FLAGS="${PBUILD_FLAGS} UNAME_r=${UNAME_r}"
		PBUILD_FLAGS="${PBUILD_FLAGS} OSREL=${REVISION}"
		PBUILD_FLAGS="${PBUILD_FLAGS} WRKDIRPREFIX=/tmp/ports"
		PBUILD_FLAGS="${PBUILD_FLAGS} DISTDIR=/tmp/distfiles"
		for _PORT in ${EMBEDDEDPORTS}; do
			eval chroot ${CHROOTDIR} env ${PBUILD_FLAGS} make -C \
				/usr/ports/${_PORT} \
				FORCE_PKG_REGISTER=1 deinstall install clean distclean
		done
	fi

	buildenv_setup

	return 0
} # extra_chroot_setup()

# chroot_build_target(): Build the userland and kernel for the build target.
chroot_build_target() {
	load_target_env
	if [ -n "${EMBEDDEDBUILD}" ]; then
		RELEASE_WMAKEFLAGS="${RELEASE_WMAKEFLAGS} \
			TARGET=${EMBEDDED_TARGET} \
			TARGET_ARCH=${EMBEDDED_TARGET_ARCH}"
		RELEASE_KMAKEFLAGS="${RELEASE_KMAKEFLAGS} \
			TARGET=${EMBEDDED_TARGET} \
			TARGET_ARCH=${EMBEDDED_TARGET_ARCH}"
	fi
	eval chroot ${CHROOTDIR} make -C /usr/src ${RELEASE_WMAKEFLAGS} buildworld
	eval chroot ${CHROOTDIR} make -C /usr/src ${RELEASE_KMAKEFLAGS} buildkernel
	if [ -n "${WITH_OCIIMAGES}" ]; then
		mkdir -p ${CHROOT}/tmp/ports ${CHROOT}/tmp/distfiles
		eval chroot ${CHROOTDIR} make -C /usr/src ${RELEASE_WMAKEFLAGS} \
		    BOOTSTRAP_PKG_FROM_PORTS=YES packages
	fi

	return 0
} # chroot_build_target

# chroot_build_release(): Invoke the 'make release' target.
chroot_build_release() {
	load_target_env
	if [ -n "${WITH_VMIMAGES}" ]; then
		if [ -z "${VMFORMATS}" ]; then
			VMFORMATS="$(eval chroot ${CHROOTDIR} \
				make -C /usr/src/release -V VMFORMATS)"
		fi
		if [ -z "${VMSIZE}" ]; then
			VMSIZE="$(eval chroot ${CHROOTDIR} \
				make -C /usr/src/release ${ARCH_FLAGS} -V VMSIZE)"
		fi
		RELEASE_RMAKEFLAGS="${RELEASE_RMAKEFLAGS} \
			VMFORMATS=\"${VMFORMATS}\" VMSIZE=${VMSIZE}"
	fi
	eval chroot ${CHROOTDIR} make -C /usr/src/release \
		${RELEASE_RMAKEFLAGS} release
	eval chroot ${CHROOTDIR} make -C /usr/src/release \
		${RELEASE_RMAKEFLAGS} install DESTDIR=/R \
		WITH_COMPRESSED_IMAGES=${WITH_COMPRESSED_IMAGES} \
		WITH_COMPRESSED_VMIMAGES=${WITH_COMPRESSED_VMIMAGES}

	return 0
} # chroot_build_release()

efi_boot_name()
{
	case $1 in
		arm)
			echo "bootarm.efi"
			;;
		arm64)
			echo "bootaa64.efi"
			;;
		amd64)
			echo "bootx64.efi"
			;;
		riscv)
			echo "bootriscv64.efi"
			;;
	esac
}

# chroot_arm_build_release(): Create arm SD card image.
chroot_arm_build_release() {
	load_target_env
	case ${EMBEDDED_TARGET} in
		arm|arm64|riscv)
			if [ -e "${RELENGDIR}/tools/arm.subr" ]; then
				. "${RELENGDIR}/tools/arm.subr"
			fi
			;;
		*)
			;;
	esac
	[ -n "${RELEASECONF}" ] && . "${RELEASECONF}"
	export MAKE_FLAGS="${MAKE_FLAGS} TARGET=${EMBEDDED_TARGET}"
	export MAKE_FLAGS="${MAKE_FLAGS} TARGET_ARCH=${EMBEDDED_TARGET_ARCH}"
	export MAKE_FLAGS="${MAKE_FLAGS} ${CONF_FILES}"
	eval chroot ${CHROOTDIR} env WITH_UNIFIED_OBJDIR=1 make ${MAKE_FLAGS} -C /usr/src/release obj
	export WORLDDIR="$(eval chroot ${CHROOTDIR} make ${MAKE_FLAGS} -C /usr/src/release -V WORLDDIR)"
	export OBJDIR="$(eval chroot ${CHROOTDIR} env WITH_UNIFIED_OBJDIR=1 make ${MAKE_FLAGS} -C /usr/src/release -V .OBJDIR)"
	export DESTDIR="${OBJDIR}/${KERNEL}"
	export IMGBASE="${CHROOTDIR}/${OBJDIR}/${BOARDNAME}.img"
	export OSRELEASE="$(eval chroot ${CHROOTDIR} make ${MAKE_FLAGS} -C /usr/src/release \
		TARGET=${EMBEDDED_TARGET} TARGET_ARCH=${EMBEDDED_TARGET_ARCH} \
		-V OSRELEASE)"
	chroot ${CHROOTDIR} mkdir -p ${DESTDIR}
	chroot ${CHROOTDIR} truncate -s ${IMAGE_SIZE} ${IMGBASE##${CHROOTDIR}}
	export mddev=$(chroot ${CHROOTDIR} \
		mdconfig -f ${IMGBASE##${CHROOTDIR}} ${MD_ARGS})
	arm_create_disk
	arm_install_base
	arm_install_boot
	arm_install_uboot
	mdconfig -d -u ${mddev}
	chroot ${CHROOTDIR} rmdir ${DESTDIR}
	mv ${IMGBASE} ${CHROOTDIR}/${OBJDIR}/${OSRELEASE}-${BOARDNAME}.img
	chroot ${CHROOTDIR} mkdir -p /R
	chroot ${CHROOTDIR} cp -p ${OBJDIR}/${OSRELEASE}-${BOARDNAME}.img \
		/R/${OSRELEASE}-${BOARDNAME}.img
	chroot ${CHROOTDIR} xz -T ${XZ_THREADS} /R/${OSRELEASE}-${BOARDNAME}.img
	cd ${CHROOTDIR}/R && sha512 ${OSRELEASE}* \
		> CHECKSUM.SHA512
	cd ${CHROOTDIR}/R && sha256 ${OSRELEASE}* \
		> CHECKSUM.SHA256

	return 0
} # chroot_arm_build_release()

# chroot_cleanup(): Clean up resources setup in chroot_setup() at exit.
#
# This contains steps which must be executed at exit.
#
# Do not override this function: override `chroot_cleanup instead.
_chroot_cleanup() {
	if [ -c "${CHROOTDIR}/dev/null" ]; then
		echo "Unmounting /dev in ${CHROOTDIR}"
		umount -f "${CHROOTDIR}/dev"
	fi
}

# main(): Start here.
main() {
	set -e # Everything must succeed
	env_setup
	while getopts c: opt; do
		case ${opt} in
			c)
				RELEASECONF="$(realpath ${OPTARG})"
				;;
			\?)
				usage
				;;
		esac
	done
	shift $(($OPTIND - 1))
	if [ -n "${RELEASECONF}" ]; then
		if [ -e "${RELEASECONF}" ]; then
			. ${RELEASECONF}
		else
			echo "Nonexistent configuration file: ${RELEASECONF}"
			echo "Using default build environment."
		fi
	fi
	env_check
	trap chroot_cleanup INT EXIT TERM
	chroot_setup
	extra_chroot_setup
	chroot_build_target
	${chroot_build_release_cmd}

	return 0
} # main()

main "${@}"
